core/storage/app_feedback_report/gae_models.py

```diff
@@ -0,0 +1,659 @@
+                web-specific feedback report.
+
+        Returns:
+            AppFeedbackReportModel. The newly created AppFeedbackReportModel
+            instance.
+        """
+        entity_id = cls._generate_id(platform, submitted_on.second)
+        android_schema_version = None
+        web_schema_version = None
+        if platform == PLATFORM_CHOICE_ANDROID:
+            android_schema_version = (
+                feconf.CURRENT_APP_FEEDBACK_REPORT_ANDROID_SCHEMA_VERSION)
+        else:
+            web_schema_version = (
+                feconf.CURRENT_APP_FEEDBACK_REPORT_WEB_SCHEMA_VERSION)
+        report_entity = cls(
+            id=entity_id, platform=platform, submitted_on=submitted_on,
+            report_type=report_type, category=category,
+            platform_version=platform_version,
+            device_country_locale_code=device_country_locale_code,
+            android_sdk_version=android_sdk_version,
+            android_device_model=android_device_model, entry_point=entry_point,
+            entry_point_topic_id=entry_point_topic_id,
+            entry_point_exploration_id=entry_point_exploration_id,
+            entry_point_story_id=entry_point_story_id,
+            entry_point_subtopic_id=entry_point_subtopic_id,
+            text_language_code=text_language_code,
+            audio_language_code=audio_language_code,
+            android_report_info=android_report_info,
+            android_report_info_schema_version=android_schema_version,
+            web_report_info=web_report_info,
+            web_report_info_schema_version=web_schema_version)
+        report_entity.update_timestamps()
+        report_entity.put()
+        return entity_id
+
+    @classmethod
+    def _generate_id(cls, platform, submitted_on_sec):
+        """Generates key for the instance of AppFeedbackReportModel class in the
+        required format with the arguments provided.
+
+        Args:
+            platform: str. The platform the user is the report from.
+            submitted_on_sec: float. The timestamp that the report was submitted
+                on, in seconds since epoch (in UTC).
+
+        Returns:
+            str. The generated ID for this entity using platform,
+            submitted_on_sec, and a random string, of the form
+            '[platform].[submitted_on_sec].[random hash]'.
+        """
+        for _ in python_utils.RANGE(base_models.MAX_RETRIES):
+            random_hash = utils.convert_to_hash(
+                python_utils.UNICODE(
+                    utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+            new_id = '%s.%s.%s' % (platform, submitted_on_sec, random_hash)
+            if not cls.get_by_id(new_id):
+                return new_id
+        raise Exception(
+            'The id generator for AppFeedbackReportModel is producing too '
+            'many collisions.')
+
+    @classmethod
+    def scrub_report(cls, report_id, scrubbed_by):
```

Comment:
Why are all these methods in the storage layer? All logical operations should typically happen in the domain layer (i.e. get the report, operate on it, and save it with the new data).

Reply:
Ah, I wasn't sure what should be included as part of model layer functions or services (e.g. the create is done in the model layer).

Will move this to the domain layer